esmvalcore/preprocessor/_area.py

```diff
@@ -295,3 +302,86 @@ def extract_named_regions(cube, regions):
+    lon_coord = cube.coord(axis='X')
+    lat_coord = cube.coord(axis='Y')
+    if lon_coord.ndim == 1 and lat_coord.ndim == 1:
+        (
+            start_longitude,
+            start_latitude,
+            end_longitude,
+            end_latitude,
+        ) = geometries.bounds
+        # add a padding of one cell around the clipped cube
+        lon_bound = lon_coord.core_bounds()[0]
+        lon_step = lon_bound[1] - lon_bound[0]
+        start_longitude -= lon_step
+        end_longitude += lon_step
+        lat_bound = lat_coord.core_bounds()[0]
+        lat_step = lat_bound[1] - lat_bound[0]
+        start_latitude -= lat_step
+        end_latitude += lat_step
+        cube = extract_region(cube, start_longitude, end_longitude,
+                              start_latitude, end_latitude)
+
+    return cube
+
+
+def extract_shape(cube, shapefile, method='contains', clip=True):
+    """Extract a region defined by a shapefile.
+
+    Parameters
+    ----------
+    cube: iris.cube.Cube
+       input cube.
+    shapefile: str
+        A shapefile defining the region(s) to extract.
+    method: str
+        Select all points contained by the shape ('contains') or
+        select a single representative point ('representative').
+    clip: bool
+        Clip the resulting cube ('true') or not ('false').
+
+    Returns
+    -------
+    iris.cube.Cube
+        Cube containing the extracted region.
+
+    """
+    if method not in {'contains', 'representative'}:
+        raise ValueError(
+            "Invalid value for `method`. Choose from 'contains', ",
+            "'representative'.")
+
+    with fiona.open(shapefile) as geometries:
+        if clip:
+            cube = _clip_geometries(cube, geometries)
+        lon_coord = cube.coord(axis='X')
+        lat_coord = cube.coord(axis='Y')
+
+        lon = lon_coord.points
+        lat = lat_coord.points
+        if lon_coord.ndim == 1 and lat_coord.ndim == 1:
+            lon, lat = np.meshgrid(lon.flat, lat.flat, copy=False)
+
+        selection = np.zeros(lat.shape, dtype=bool)
+        for item in geometries:
+            shape = shapely.geometry.shape(item['geometry'])
+            if method == 'contains':
+                select = shapely.vectorized.contains(shape, lon, lat)
+            if method == 'representative' or not select.any():
+                representative_point = shape.representative_point()
+                points = shapely.geometry.MultiPoint(
+                    np.stack((lon.flat, lat.flat), axis=1))
+                nearest_point = shapely.ops.nearest_points(
+                    points, representative_point)[0]
+                nearest_lon, nearest_lat = nearest_point.coords[0]
+                select = (lon == nearest_lon) & (lat == nearest_lat)
+            selection |= select
+
+        selection = da.broadcast_to(selection, cube.shape)
+        cube.data = da.ma.masked_where(~selection, cube.core_data())
+        return cube
```

Comment:
it'd probably useful to add a reference in the `cube.metadata` something like `contains whatever region only`

Reply:
I'm not sure if that is needed, `extract_region` does not do that either

Reply:
yeah by extension that's be useful there too

Reply:
Implementing this is also not trivial, because there is no descriptive name available. There is just a shapefile.

Reply:
Maybe adding a generic "masked irregular shape" to the metadata would work so it can be discerned from a cube of the same size (for example obtained with `extract_region`) without looking at the data.

Reply:
I've thought about this some more and I think it's not a good idea. We have provenance for recording what happened to exactly to the data and it does not make sense to start recording some of it also as cube attributes.